requestmanager/client.go

```diff
@@ -0,0 +1,359 @@
+package requestmanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/hannahhoward/go-pubsub"
+	blocks "github.com/ipfs/go-block-format"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipfs/go-graphsync/listeners"
+	gsmsg "github.com/ipfs/go-graphsync/message"
+	"github.com/ipfs/go-graphsync/messagequeue"
+	"github.com/ipfs/go-graphsync/metadata"
+	"github.com/ipfs/go-graphsync/notifications"
+	"github.com/ipfs/go-graphsync/requestmanager/hooks"
+	"github.com/ipfs/go-graphsync/requestmanager/types"
+)
+
+// The code in this file implements the public interface of the request manager.
+// Functions in this file operate outside the internal thread and should
+// NOT modify the internal state of the RequestManager.
+
+var log = logging.Logger("graphsync")
+
+const (
+	// defaultPriority is the default priority for requests sent by graphsync
+	defaultPriority = graphsync.Priority(0)
+)
+
+type inProgressRequestStatus struct {
+	ctx            context.Context
+	startTime      time.Time
+	cancelFn       func()
+	p              peer.ID
+	terminalError  chan error
+	resumeMessages chan []graphsync.ExtensionData
+	pauseMessages  chan struct{}
+	paused         bool
+	lastResponse   atomic.Value
+	onTerminated   []chan<- error
+}
+
+// PeerHandler is an interface that can send requests to peers
+type PeerHandler interface {
+	AllocateAndBuildMessage(p peer.ID, blkSize uint64, buildMessageFn func(*gsmsg.Builder), notifees []notifications.Notifee)
+}
+
+// AsyncLoader is an interface for loading links asynchronously, returning
+// results as new responses are processed
+type AsyncLoader interface {
+	StartRequest(graphsync.RequestID, string) error
+	ProcessResponse(responses map[graphsync.RequestID]metadata.Metadata,
+		blks []blocks.Block)
+	AsyncLoad(p peer.ID, requestID graphsync.RequestID, link ipld.Link, linkContext ipld.LinkContext) <-chan types.AsyncLoadResult
+	CompleteResponsesFor(requestID graphsync.RequestID)
+	CleanupRequest(requestID graphsync.RequestID)
+}
+
+// RequestManager tracks outgoing requests and processes incoming reponses
+// to them.
+type RequestManager struct {
+	ctx             context.Context
+	cancel          func()
```

Comment:
`context.CancelFunc`?

Reply:
currently it's just func() everywhere in the code and I should change it all at once. will do in seperate cleanup PR